calico_containers/libnetwork_plugin/docker_plugin.py

```diff
@@ -216,13 +199,45 @@ def endpoint_oper_info():
 @app.route('/NetworkDriver.Join', methods=['POST'])
 def join():
     json_data = request.get_json(force=True)
+    app.logger.debug("Join JSON=%s", json_data)
     ep_id = json_data["EndpointID"]
+    net_id = json_data["NetworkID"]
     app.logger.info("Joining endpoint %s", ep_id)
 
-    ep = client.get_endpoint(hostname=hostname,
-                             orchestrator_id="docker",
-                             workload_id=CONTAINER_NAME,
-                             endpoint_id=ep_id)
+    # Get CNM endpoint ID from datastore so we can find the IP addresses
+    # assigned to it.
+    cnm_ep = client.read_cnm_endpoint(ep_id)
+
+    # Read the next hops from etcd
+    next_hops = client.get_default_next_hops(hostname)
+
+    # Create a Calico endpoint object.
+    #TODO - set the CONTAINER_NAME to something better (the sandbox key?)
+    ep = Endpoint(hostname, "docker", CONTAINER_NAME, ep_id, "active",
+                      FIXED_MAC)
+    ep.profile_ids.append(net_id)
+    ep.ipv4_nets.add(IPNetwork(cnm_ep['Interfaces'][0]['Address']))
+    ep.ipv4_gateway = next_hops[4]
+
+    address_ip6 = cnm_ep['Interfaces'][0].get('AddressIPv6')
+    app.logger.info("Checking IPv6 %s", address_ip6)
+    if address_ip6:
+        ep.ipv4_nets.add(IPNetwork(address_ip6))
+        ep.ipv4_gateway = next_hops[6]
+
+    try:
+        # Next, create the veth.
+        create_veth(ep)
+
+        # Finally, write the endpoint to the datastore.
+        client.set_endpoint(ep)
+    except (CalledProcessError, DataStoreError) as e:
```

Comment:
Combining these two calls into a single try seems a bit skanky to me since they both raise different exceptions, and I'd expect if create_veth failed that it wouldn't be necessary to call remove_veth.